Add contact preview component and use it on contact embed insertion

alex-ju opened the pull request:
This PR:
- creates a new contact preview component (following the pattern used in the url preview component) - this component refers and uses a select element as a source of truth, a govspeak snippet template and a govspeak path (all being passed as data attributes at component's root level)
- uses the contact preview component it in the contact embed view and
- updates the contact autocomplete to properly sync the select element that's enhanced by it.

See previews in the [previous approach](https://github.com/alphagov/content-publisher/pull/1310) (as the outcome is the same).

[Trello card](https://trello.com/c/8NN6jQiI/1025-modal-experience-for-contacts)
[Trello card](https://trello.com/c/6uzRis6D/1145-when-a-user-chooses-a-contact-they-can-preview-it-and-insert-it-in-content)

alex-ju commented on app/assets/javascripts/components/contact-preview.js:
```diff
@@ -0,0 +1,103 @@
+    if (this.selectId) {
+      this.$select = document.querySelector('#' + this.selectId)
+      if (this.$select) {
+        this.contactSnippetTemplate = this.$module.dataset.contactSnippetTemplate
```
Updated

alex-ju commented on app/assets/javascripts/components/contact-preview.js:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  ContactPreview.prototype.getContactSnippet = function () {
+    if (this.$select && this.contactSnippetTemplate && this.$select.value) {
```
Fair, we should have it by then. Updated.

alex-ju commented on app/assets/javascripts/components/contact-preview.js:
```diff
@@ -0,0 +1,103 @@
+    this.errorMessage = this.$module.querySelector('.js-contact-preview-error-message')
+    this.contactPreview = this.$module.querySelector('.js-contact-preview-html')
+    this.selectId = this.$module.dataset.for
+    if (this.selectId) {
```
We can rely on the inner one, since it won't blow if the `querySelector` looks for an invalid id. Updated.

alex-ju commented on app/assets/javascripts/components/contact-preview.js:
```diff
@@ -0,0 +1,103 @@
+    var url = new URL(document.location.origin + this.path)
+
+    var formData = new window.FormData()
+    var contactSnippet = this.getContactSnippet(contactId)
```
Yep, this parameter is not needed anymore.

alex-ju commented on app/views/components/_contact_preview.html.erb:
```diff
@@ -0,0 +1,17 @@
+%>
+<%= tag.div class: root_classes, data: data_attributes.merge(module: "contact-preview") do %>
+  <% contact_preview = capture do %>
+    <%= tag.p error_message, class: "app-c-contact-preview__error-message app-c-contact-preview__error-message--hidden js-contact-preview-error-message" %>
```
We use the `js-` classes to execute query selectors - sort of a way of saying this class is not used for styling but selection.

benthorner replied:
👍

alex-ju commented on spec/javascripts/components/contact-preview-spec.js:
```diff
@@ -0,0 +1,74 @@
+
+    var previewContainer = document.querySelector('.js-contact-preview-html')
+    expect(previewContainer).toBeVisible()
+    expect(contactPreview).toEqual(contactPreview)
```
😄It should've been `previewContainer.innerHtml`. Updated.

alex-ju commented on spec/javascripts/components/contact-preview-spec.js:
```diff
@@ -0,0 +1,74 @@
+    expect(errorMessage).toBeHidden()
+  })
+
+  it('should remove the preview container if no option is selected', async function () {
```
Yep, updated.

alex-ju commented on app/views/components/docs/contact_preview.yml:
```diff
@@ -0,0 +1,25 @@
+name: Contact preview
+description: Generates a preview of a given govspeak contact snippet from a select element
+body: |
+  This component uses a select element to get the contact ID, converts the ID to a govspeak contact snippet
```
Updated to capture the static approach too.

benthorner commented:
@kevindew since I've now contributed to this, would you mind taking over the review?

kevindew commented:
I just gave this a spin at my end to start a review and I've caught a bug.

If you change your contact two preview requests seem to fire and your new contact you've selected gets replaced by the old one:

![Aug-09-2019 18-07-34](https://user-images.githubusercontent.com/282717/62796238-a8d4e800-bad0-11e9-85f1-b6bdbd8f8c11.gif)

Presumably this is some consequence of the rather odd onConfirm behaviour that accessible-autocomplete can have in some way. In my example above it seems to be a consequence of losing focus.

In my testing the loading of a contact is quite slow ~1.5s and it showed a couple of issues:

1) It'd be nice if it showed a waiting indication when the preview is about to appear. The jump to a new contact is a bit jarring
2) It's a shame that out of date isn't replaced until a new contact is available.
3) Every time you lose focus on autocomplete a new HTTP request is sent to update the preview even if nothing changed

I imagine we also need something more in the way of aborting when a user changes selection before something is loaded.

alex-ju commented:
@kevindew thanks for the review. All valid concerns from my point of view. It's a shame we've spent the sprint trying to find the best way to communicate with the backend instead of improving the interaction. I can leave this PR open, but I'm starting to question it if no-one will be able to work on it for a while.

leenagupte commented:
@kevindew has confirmed that it is ok to close this PR as no work is going to be done on it for a while. It can be re-opened when needed.

alex-ju commented:
@kevindew tried to address your comments with the last two commits.

> * It'd be nice if it showed a waiting indication when the preview is about to appear. The jump to a new contact is a bit jarring

Added a loading spinner while we're waiting for fetch to return a result in 3ec085b

> * It's a shame that out of date isn't replaced until a new contact is available.

Now after selecting a new option (sending a new fetch request) we're replacing the old contact with the loading spinner. If the input is cleared we remove the existing contact.

> * Every time you lose focus on autocomplete a new HTTP request is sent to update the preview even if nothing changed

Fixed this in 606d397 which also addresses the bug you've found — if I got it right.

leenagupte commented on app/controllers/contact_embed_controller.rb:
```diff
@@ -7,24 +7,30 @@ class ContactEmbedController < ApplicationController
     result = ContactEmbed::CreateInteractor.call(params: params)
-    markdown_code, issues = result.to_h.values_at(:markdown_code, :issues)
+
+    edition, markdown_code, issues = result.to_h.values_at(:edition,
```
I'm not sure why `edition` needs to be returned by `ContactEmbed::CreateInteractor`. Is it possible to do the same thing as `new` and call `@edition = Edition.find_current(document: params[:document])`?

alex-ju replied:
`edition` is required because the govspeak snipped for contacts require an edition in order to rendered.

It may be possible, but by doing it this way we are consistent with how we do it for [images](https://github.com/alphagov/content-publisher/blob/master/app/controllers/images_controller.rb#L12) and [file attachments](https://github.com/alphagov/content-publisher/blob/master/app/controllers/file_attachments_controller.rb#L25) — not really my call to change this pattern.

leenagupte replied:
I had a look at those examples, and they lock the edition as well, which doesn't seem to be needed for inserting a contact?

alex-ju replied:
As far as I got it back then (3 months ago) we needed to lock it for the contacts too. I'll be sneaky and @benthorner on this — as he has more background on it than I do.

alex-ju commented:
Thank you for reviewing this @leenagupte! I will merge it as is and it's on me to update the controller if Ben or Kevin suggest otherwise.
